fetchcode/package.py

```diff
@@ -0,0 +1,318 @@
+            code_view_url=code_view_url,
+            download_url=download_url,
+            declared_license=declared_license,
+            **vpurl.to_dict(),
+        )
+
+
+@router.route("pkg:npm/.*")
+def get_npm_data_from_purl(purl):
+    """
+    Generate `Package` object from the `purl` string of npm type
+    """
+    purl = PackageURL.from_string(purl)
+    base_path = "http://registry.npmjs.org"
+    name = purl.name
+    version = purl.version
+    api_url = f"{base_path}/{name}"
+    response = get_response(api_url)
+    vcs_data = response.get("repository") or {}
+    bugs = response.get("bugs") or {}
+
+    download_url = f"{base_path}/{name}/-/{name}-{version}.tgz"
+    vcs_url = vcs_data.get("url")
+    bug_tracking_url = bugs.get("url")
+    license = response.get("license")
+    homepage_url = response.get("homepage")
+
+    yield Package(
+        homepage_url=homepage_url,
+        api_url=api_url,
+        vcs_url=vcs_url,
+        bug_tracking_url=bug_tracking_url,
+        download_url=download_url,
+        declared_license=license,
+        **purl.to_dict(),
+    )
+
+    versions = response.get("versions", [])
+    tags = []
+    for num in versions:
+        version = versions[num]
+        vpurl = PackageURL(type=purl.type, name=name, version=version.get("version"))
+        repository = version.get("repository") or {}
+        bugs = response.get("bugs") or {}
+        dist = version.get("dist") or {}
+        licenses = version.get("licenses") or [{}]
+        vcs_url = repository.get("url")
+        download_url = dist.get("tarball")
+        bug_tracking_url = bugs.get("url")
+        declared_license = licenses[0].get("type")
+
+        yield Package(
+            homepage_url=homepage_url,
+            api_url=api_url,
+            vcs_url=vcs_url,
+            bug_tracking_url=bug_tracking_url,
+            download_url=download_url,
+            declared_license=declared_license,
+            **vpurl.to_dict(),
+        )
+
+
+@router.route("pkg:pypi/.*")
+def get_pypi_data_from_purl(purl):
+    """
+    Generate `Package` object from the `purl` string of npm type
+    """
+    purl = PackageURL.from_string(purl)
+    name = purl.name
+    base_path = "https://pypi.org/pypi"
+    api_url = f"{base_path}/{name}/json"
+    response = get_response(api_url)
+    releases = response.get("releases") or {}
+    info = response.get("info") or {}
+    homepage_url = info.get("home_page")
+    license = info.get("license")
+    project_urls = info.get("project_urls") or {}
+    code_view_url = project_urls.get("Source")
+    bug_tracking_url = project_urls.get("Tracker")
+    if not (code_view_url):
+        code_view_url = project_urls.get("Code")
+    if not (bug_tracking_url):
+        bug_tracking_url = project_urls.get("Issue Tracker")
+    if not (code_view_url):
+        code_view_url = project_urls.get("Source Code")
+    if not (bug_tracking_url):
+        bug_tracking_url = project_urls.get("Bug Tracker")
```

Comment:
The logic here is confusing. We probably want seperate functions for these if there are multiple steps needed to construct these urls.

Reply:
These are not multiple steps, Pypi does not have consistent keys, so I check for every name that the key can have.

Reply:
I am saying that logic should be in a separate function

Reply:
You are right, makes sense to me too

Reply:
Done :)